### PR TITLE
Start simulation with neutral atoms

### PR DESCRIPTION
**Pull request by n01r**
It will be possible from now on to start the simulation neutrally with un-ionized atoms. 
Therefore the number of electrons bound to the core is set to the number of protons of the ion species via the `manipulator` concept during initialization. 

Furthermore ionization energies and the fields necessary to overcome them are now stored in `PMACC_CONST_VECTOR()` arrays and can be accessed by the ionization algorithm according to the current charge state. 

This allows now for any chemical element to be implemented. With it also comes *for free* that the effective charge of the initialized atoms is equal to zero because it is calculated from the `numberOfProtons` and the number of `boundElectrons`.

If you happen to implement different elements as species, remember to change the weight of the atom / ion accordingly. A more generic approach should follow in due time.

2015_03_23_bsi_electrons

In the picture above a simulation result of Hydrogen atoms being ionized can be seen (electron density is depicted). I guess that these funny "filament-structures" are being caused by the step-like nature of the BSI algorithm.

### CODE REVIEW

**Comment by PrometheusPi** [comment]
great job :+1:

**Comment by bussmann** [comment]
Like it!

**Comment by n01r** [comment]
OK, to clarify things a bit. I did not pay so much attention, yet as to how the simulation parameters fit to each other. The ionization model still uses only the **z - component** of the electric field and the laser was **circularly polarized**.

I now tried a bigger simulation area and a **linear polarization in z direction**. 
animation_marco

The structures are still there. I could change the sign of the amplitude of the laser field to see if the regions of high and low density get mirrored. (Btw: 5fps look horrible :stuck_out_tongue: ... but a 42,9MB GIF was no viable option)

**Comment by ax3l** [comment]
well done @n01r, let's review the changes! :)

File: examples/LaserWakefield/cmakeFlags [hunk 1]
```diff
@@ -41,7 +41,7 @@ flags[8]="-DCUDA_ARCH=sm_20 -DPARAM_OVERWRITES:LIST=-DPARAM_DIMENSION=DIM2"
 flags[9]="-DCUDA_ARCH=sm_20 -DPARAM_OVERWRITES:LIST=-DPARAM_CURRENTSOLVER=Esirkepov;-DPARAM_PARTICLESHAPE=CIC;-DPARAM_DIMENSION=DIM2"
 flags[10]="-DCUDA_ARCH=sm_20 -DPARAM_OVERWRITES:LIST=-DPARAM_PRECISION=precision64Bit;-DPARAM_DIMENSION=DIM2"
 flags[11]="-DCUDA_ARCH=sm_35 -DPARAM_OVERWRITES:LIST=-DENABLE_CURRENT=0;-DPARAM_DIMENSION=DIM2"
-flags[12]="-DCUDA_ARCH=sm_35 -DPARAM_OVERWRITES:LIST=-DPARAM_GASPROFILE=gasSphereFlanks;-DPARAM_IONS=1;-DPARAM_IONIZATION=1"
+flags[12]="-DCUDA_ARCH=sm_35 -DPARAM_OVERWRITES:LIST=-DPARAM_IONS=1;-DPARAM_IONIZATION=1"
```
Comment (ax3l) [inline examples/LaserWakefield/cmakeFlags]: hm, why did you change that? do you mind reverting that change?

File: src/picongpu/include/particles/ionization/ionizationEnergies.param [hunk 2]
```diff
@@ -36,9 +36,23 @@ namespace SI
 
     /* ionization energy for ground state hydrogen: 13.6 eV --> converted to Joule */
     /* Joule = kg * m^2 / s^2 */
-    const double IONIZATION_ENERGY = 2.179e-18; 
-    /* ionization field strength (E-field) for -""-: in Volt / meter */
-    const double IONIZATION_EFIELD = 5.14e11;
+    PMACC_CONST_VECTOR(float_X,1,IONIZATION_ENERGY,
+        2.179e-18
+    );
+    /* ionization field strength (E-field) for -""-: in Volt / meter
```
Comment (ax3l) [inline src/picongpu/include/particles/ionization/ionizationEnergies.param]: what is `-""-` ?

File: src/picongpu/include/particles/ionization/ionizationEnergies.param [hunk 3]
```diff
@@ -36,9 +36,23 @@ namespace SI
...
+     * usage: 
+     * PMACC_CONST_VECTOR(float_X,<NUMBER_OF_PROTONS>,IONIZATION_EFIELD,
+     *     <value_1>,
+     *     ...
+     *     <value_Z>
+     * );
+     * 
+     * Do not forget to change the weight of the ion and the atomic numbers in 
+     * their respective `.param` file */
+    PMACC_CONST_VECTOR(float_X,1,IONIZATION_EFIELD,
```
Comment (ax3l) [inline src/picongpu/include/particles/ionization/ionizationEnergies.param]: what do I do if I have two ion species in my simulation that can both be ionized?

Comment (ax3l) [inline src/picongpu/include/particles/ionization/ionizationEnergies.param]: (can be answered by opening an issue for that; feel free to address that in a follow up pull request)

**Comment by ax3l** [comment]
looks great to me! :sparkles: 

just some inline questions left.

File: examples/LaserWakefield/cmakeFlags (hunk 1 above)
Comment (n01r) [inline examples/LaserWakefield/cmakeFlags]: I removed it because @psychocoderHPC 's big pull request #730 removed all of the gas configuration via cmake flags. Test case 12 was originally introduced by me anyway. I plan to make gas droplet studies later but for now this `-DPARAM_GASPROFILE=gasSphereFlanks` was totally without effect for all of the structure was changed.

File: src/picongpu/include/particles/ionization/ionizationEnergies.param (hunk 2 above)
Comment (n01r) [inline src/picongpu/include/particles/ionization/ionizationEnergies.param]: it's a happy 6-armed creature ... just kidding, my bad :stuck_out_tongue: 
the name of an element from the PSE is supposed to be here

File: src/picongpu/include/particles/ionization/ionizationEnergies.param (hunk 3 above)
Comment (n01r) [inline src/picongpu/include/particles/ionization/ionizationEnergies.param]: Yep ... it would be possible but very inelegant to just add another differently named array of ionization energies and use them. Therefore let's think about it together in #771 .

File: examples/LaserWakefield/cmakeFlags (hunk 1 above)
Comment (ax3l) [inline examples/LaserWakefield/cmakeFlags]: ah true, thanks!

**Comment by ax3l** [comment]
great work, well done! :+1:

**Comment by n01r** [comment]
Yay that was fast :fast_forward:
